src/persist/src/indexed/mod.rs

```diff
@@ -689,11 +714,93 @@ impl<L: Log, B: Blob> Indexed<L, B> {
             self.blob.delete_unsealed_batch(&batch)?;
         }
 
-        for batch in deleted_trace_batches {
-            self.blob.delete_trace_batch(&batch)?;
+        Ok(())
+    }
+
+    fn compact_trace_maintenance_reqs(&mut self) -> Result<Vec<MaintenanceReq>, Error> {
+        let mut reqs = vec![];
+        for (id, arrangement) in self.state.arrangements.iter() {
+            // Skip over the arrangement if it already has a compaction request in flight.
+            //
+            // TODO: note that this works because right now the connection between Indexed
+            // and Maintainer is infallible, and the two cannot fail independently of each
+            // other. If that changes, we will have to revisit the policy around waiting
+            // arbitrarily long for maintenance responses and most likely institute
+            // a timeout.
+            if self.in_flight_trace_compactions.contains_key(&id) {
+                continue;
+            }
+            if let Some(req) = arrangement.trace_next_compact_req()? {
+                reqs.push(MaintenanceReq::CompactTrace((*id, req.clone())));
+                self.in_flight_trace_compactions.insert(*id, req);
+            }
+        }
+        Ok(reqs)
+    }
+
+    fn do_handle_compact_trace_res(&mut self, id: Id, res: CompactTraceRes) {
+        let written_bytes = res.merged.size_bytes;
+        self.metrics.compaction_write_bytes.inc_by(written_bytes);
+
+        if !self.state.arrangements.contains_key(&id) {
+            log::trace!(
+                "received trace compaction response for deleted stream id: {:?}. Ignoring.",
+                id
+            );
+            return;
         }
 
-        Ok(())
+        self.apply_batched_cmd(|state, pending| {
+            let arrangement = state
+                .arrangements
+                .get_mut(&id)
+                .expect("arrangement known to exist");
+            let (_, deleted_trace_batches) = arrangement.trace_handle_compact_response(res);
+            if !deleted_trace_batches.is_empty() {
+                pending.add_deleted_trace_batches(id, deleted_trace_batches);
+            }
+        });
+    }
+
+    /// Handle the results of a previously sent maintenance request.
+    fn handle_maintenance_res(&mut self, res: MaintenanceRes) {
+        match res {
+            MaintenanceRes::CompactTrace((id, res)) => {
+                if !self.in_flight_trace_compactions.contains_key(&id) {
+                    log::trace!(
+                        "received trace compaction response for a stream that did not have a in flight request: {:?}. Ignoring",
+                        id
+                    );
+                    return;
+                }
+
+                match res {
+                    Ok(res) => {
+                        if let Some(sent_req) = self.in_flight_trace_compactions.get(&id) {
+                            if res.req != *sent_req {
+                                log::trace!(
```

Comment:
i don't love this != check (it feels like a smell to me, but I can't put my finger on why) and still have a mild preference for simply handing them. i didn't particularly find your counter-argument compelling: that's what unit tests are for

however, it isn't a strong preference, so if we're going to commit to not handling them, we should add a TODO so the thought doesn't get lost and it should assert in debug (and possibly log in prod but honestly probably just ignore them entirely and return). that way it actually gets noticed if it happens while we're running tests. a unactionable log is kinda the worst of all worlds

ditto for the other log::trace above

Reply:
Let me try again with the counterargument.

The scenario where we receive an unexpected MaintenanceRes really only makes sense to me in the context of "Indexed and Maintainer are separate processes communicating over a fallible network". Otherwise, as you say, its a bug in Maintainer and we have unit testing to shake that out.

If there is a fallible network i guess my concern is that:

Indexed sends 1 request to Maintainer
Maintainer sends just the response to that request 100 times and doesn't respond to any other request (perhaps because of some transient error)

in that world, I was worried about Indexed sending 99 extra requests after the unnecessary 99 responses. Ideally, I think something like what you're saying makes sense. We just need to try to apply the response, and if its useful, great! and if its not, we ignore it, and we keep waiting for a useful response to the request that we sent (I'm assuming we don't have any random network corruption as well). Right now, because of how we're batching the responses while handling them, this seems hard to do (because `apply_batched_cmd` drops the result of its work function). This all seemed doable, but it seemed like complexity that might shoot me in the foot, whereas the != check felt like extra complexity that I could easily understand.

Reply:
I changed the log::trace to `soft_assert_or_log` which makes this harder to test because now these conditions panic, but worked around that with some elbow grease.